### ISSUE TITLE
Potential skin/bone animation bug

### ISSUE THREAD

**Issue by Basewq** [stride3d/stride#2284]
# Question
Need research into the behavior of data transfer between a `VertexElement` declaration and the data type declared in a shader.
I noticed this bug on Android testing the Top-down RPG project, where the player was not animating.

https://github.com/stride3d/stride/blob/1fc0d5137750a8c3d8acd5f89e3b63c75b6bea7f/sources/engine/Stride.Rendering/Rendering/Skinning/TransformationSkinning.sdsl#L28-L33

Notice that 'BLENDINDICES' is `uint4`, but `float4` for ES2 (which I think hints at the bug).

However the `MeshConverter` changes the `VertexElement` data type of the BLENDINDICES depends on the *mesh's* settings:
https://github.com/stride3d/stride/blob/1fc0d5137750a8c3d8acd5f89e3b63c75b6bea7f/sources/tools/Stride.Importer.3D/MeshConverter.cs#L724-L753

Does BLENDINDICES's data get transferred correctly with the AllowUnsignedBlendIndices/16bit indices settings on the mesh? ie. will it cast to match the shader's data type, or will it just mismatch silently and be wrong?

**Comment by Basewq** [comment]
The following answer is based on the [answer here on stackoverflow](https://stackoverflow.com/a/34013624) (which doesn't appear to be any official documentation on the behavior), and through my own testing and checking through RenderDoc:

TLDR: Yes, there is a bug in the shader code, and mismatching data types are problematic.

Longer answer: The shader is expected to declare a data type compatible to that declared in the `VertexElement`, therefore incorrect data type will break in the shader (ie. Android/OpenGL ES is 100% broken).

I choose the word "compatible", because shader data type is not exactly the same as C# data type.
The important thing to check is the suffix of the name of the `PixelFormat` type (eg. `_UInt`, `_SInt`), not the bit count (eg. `R8` vs `R16`).
eg. 8 bit uint & 16 bit uint will be ok for a shader to declare `uint`. But declaring `float` will be wrong due to different bit representation, `int` will show issues when the bit interpretation changes (ie. negative value/value larger than int.max) - better to not make assumptions here and always use the right data type.
In summary:
* `_FLOAT`, `_UNORM`, `_SNORM` should use `float`
* `_SInt` should use `int`
* `_UInt` should use `uint`

Appropriate fix for the shader is changing the macro for signed vs unsigned blend indices.
The top note in the shader code:
https://github.com/stride3d/stride/blob/1fc0d5137750a8c3d8acd5f89e3b63c75b6bea7f/sources/engine/Stride.Rendering/Rendering/Skinning/TransformationSkinning.sdsl#L8
is probably no longer applicable due to ES 2 already dropped.
`uint4 BlendIndices` does not appear to have an issue on Android with ES 3, but may need to double check this again.
